### PR TITLE
Implement OpenSSL Engine tests for NPN / ALPN.

### PR DESCRIPTION
**Pull request by ifesdjeen**
Motivation: 

Currently there are no tests for OpenSSL Engine, only for JdkSSL engine.

Modifications: 

Common methods from `JdkSslEngine` test moved to `SSLEngineTest`, JdkSslEngine now implements
NPN and ALPN tests. 

Result:
 
OpenSSL Engine is now covered with unit tests.

### CODE REVIEW

File: handler/src/test/java/io/netty/handler/ssl/OpenSslEngineTest.java [hunk 1]
```diff
@@ -58,4 +98,12 @@ public void testSessionInvalidate() throws Exception {
     protected SslProvider sslProvider() {
         return SslProvider.OPENSSL;
     }
+
+    private ApplicationProtocolConfig acceptingNegotiator(Protocol protocol,
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/OpenSslEngineTest.java]: static

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 2]
```diff
@@ -380,4 +381,95 @@ private static void runDelegatedTasks(SSLEngineResult result, SSLEngine engine)
...
+
+        sb.group(new NioEventLoopGroup(), new NioEventLoopGroup());
+        sb.channel(NioServerSocketChannel.class);
+        sb.childHandler(new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                ChannelPipeline p = ch.pipeline();
+                p.addLast(serverSslCtx.newHandler(ch.alloc()));
+                p.addLast(new MessageDelegatorChannelHandler(serverReceiver, serverLatch));
+                p.addLast(new ChannelHandlerAdapter() {
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: ChannelInboundHandlerAdapter please as we want to remove the exceptionCaught(...) method from ChannelHandler

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 3]
```diff
@@ -380,4 +381,95 @@ private static void runDelegatedTasks(SSLEngineResult result, SSLEngine engine)
...
+
+        cb.group(new NioEventLoopGroup());
+        cb.channel(NioSocketChannel.class);
+        cb.handler(new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                ChannelPipeline p = ch.pipeline();
+                p.addLast(clientSslCtx.newHandler(ch.alloc()));
+                p.addLast(new MessageDelegatorChannelHandler(clientReceiver, clientLatch));
+                p.addLast(new ChannelHandlerAdapter() {
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: ChannelInboundHandlerAdapter please as we want to remove the exceptionCaught(...) method from ChannelHandler

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 4]
```diff
@@ -380,4 +381,95 @@ private static void runDelegatedTasks(SSLEngineResult result, SSLEngine engine)
...
+                        }
+                    }
+                });
+            }
+        });
+
+        serverChannel = sb.bind(new InetSocketAddress(0)).sync().channel();
+        int port = ((InetSocketAddress) serverChannel.localAddress()).getPort();
+
+        ChannelFuture ccf = cb.connect(new InetSocketAddress(NetUtil.LOCALHOST, port));
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: `cb.connect(serverChannel.localAddress())` ? and remove the above line which gets the port.

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 5]
```diff
@@ -380,4 +381,95 @@ private static void runDelegatedTasks(SSLEngineResult result, SSLEngine engine)
...
+                            clientLatch.countDown();
+                        } else {
+                            ctx.fireExceptionCaught(cause);
+                        }
+                    }
+                });
+            }
+        });
+
+        serverChannel = sb.bind(new InetSocketAddress(0)).sync().channel();
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: `sync()` -> `syncUninterruptibly()`

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 6]
```diff
@@ -380,4 +381,95 @@ private static void runDelegatedTasks(SSLEngineResult result, SSLEngine engine)
...
+        int port = ((InetSocketAddress) serverChannel.localAddress()).getPort();
+
+        ChannelFuture ccf = cb.connect(new InetSocketAddress(NetUtil.LOCALHOST, port));
+        assertTrue(ccf.awaitUninterruptibly().isSuccess());
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: `awaitUninterruptibly()` -> `syncUninterruptibly()`

**Comment by ifesdjeen** [comment]
Thanks for reviewing @normanmaurer. 
I've adressed all the comments. 

It looks like CI checks aren't running because the PR is against 4.1 but not master (as far as I understand you have changed the project release strategy). I've ran the tests and checks locally though...

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 7]
```diff
@@ -380,4 +381,94 @@ private static void runDelegatedTasks(SSLEngineResult result, SSLEngine engine)
...
+            }
+        });
+
+        serverChannel = sb.bind(new InetSocketAddress(0)).syncUninterruptibly().channel();
+
+        ChannelFuture ccf = cb.connect(serverChannel.localAddress());
+        assertTrue(ccf.syncUninterruptibly().isSuccess());
+        clientChannel = ccf.channel();
+    }
+
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: nit: kill line

**Comment by Scottmitch** [comment]
Thanks @ifesdjeen! There does seem do be an issue with the CI server https://github.com/netty/netty/issues/4632. I built locally and everything checked out.

Cherry-picked 4.0 (b9dc4f4) 4.1 (ba22b0b)

**Comment by normanmaurer** [comment]
@ifesdjeen @Scottmitch hmm... seems like this fails on the CI. Can you check:

https://garage.netty.io/teamcity/viewLog.html?buildId=4308&buildTypeId=netty_build_oraclejdk7

**Comment by Scottmitch** [comment]
@normanmaurer - Looking.

**Comment by Scottmitch** [comment]
@normanmaurer - I was able to reproduce the failure locally. How I was able to reproduce was having a version of OpenSSL that did not support ALPN (so the ALPN tests would fail but so `assumeTrue(OpenSsl.isAvailable());` would allow the test to run). I will submit a PR to skip the tests unless the OpenSsl version supports ALPN but I would like to confirm the configuration of the build servers to see if this is same conditions that lead to the failure. Can you describe the build/run environment of the slave that encountered this failure?

**Comment by ifesdjeen** [comment]
Whoopsie sorry @Scottmitch @normanmaurer. That's why i wanted to get it CIed first :) Thanks for taking care of it in #4633

**Comment by Scottmitch** [comment]
@ifesdjeen - Wait for CI first ... I should have thought of that :wink: ... no worries not your fault.

**Pull request closed without merging**